### PR TITLE
Replace ADAL with MSAL, Replace methods lost in the transition to MSAL-affiliated libraries, Expire and retrieve Azure Batch Quota after 24 hours

### PR DESCRIPTION
**Pull request by BMurri**
Companion to microsoft/CromwellOnAzure#794
Addresses #744

### CODE REVIEW

File: src/deploy-tes-on-azure/Utility.cs [hunk 1]
```diff
@@ -134,5 +144,49 @@ public static string GetFileContent(params string[] pathComponentsRelativeToAppB
...
+        private class ResourceNamer
+        {
+            private readonly string randName;
+            private static readonly Random random = new();
+
+            public ResourceNamer(string name)
+            {
+                lock (random)
+                {
+                    this.randName = name.ToLower() + Guid.NewGuid().ToString("N")[..3].ToLower();
```
Comment (MattMcL4475) [inline src/deploy-tes-on-azure/Utility.cs:158]: ToLowerInvariant

File: src/deploy-tes-on-azure/Utility.cs [hunk 2]
```diff
@@ -134,5 +144,49 @@ public static string GetFileContent(params string[] pathComponentsRelativeToAppB
...
+
+        // borrowed from https://github.com/Azure/azure-libraries-for-net/blob/7d85e294e4e7280c3f74b1c41438e2f20bce2052/src/ResourceManagement/ResourceManager/ResourceNamer.cs
+        private class ResourceNamer
+        {
+            private readonly string randName;
+            private static readonly Random random = new();
+
+            public ResourceNamer(string name)
+            {
+                lock (random)
```
Comment (MattMcL4475) [inline src/deploy-tes-on-azure/Utility.cs:156]: why does this need a lock?  seems like all the locks can be removed

File: src/deploy-tes-on-azure/Utility.cs [hunk 3]
```diff
@@ -134,5 +144,49 @@ public static string GetFileContent(params string[] pathComponentsRelativeToAppB
...
+                {
+                    this.randName = name.ToLower() + Guid.NewGuid().ToString("N")[..3].ToLower();
+                }
+            }
+
+            public string RandomName(string prefix, int maxLen)
+            {
+                lock (random)
+                {
+                    prefix = prefix.ToLower();
```
Comment (MattMcL4475) [inline src/deploy-tes-on-azure/Utility.cs:166]: ToLowerInvariant

File: src/deploy-tes-on-azure/Properties/launchSettings.json [hunk 4]
```diff
@@ -2,7 +2,7 @@
   "profiles": {
     "Setup": {
       "commandName": "Project",
-      "commandLineArgs": "--subscriptionid aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee --regionname westus2 --mainidentifierprefix toatest --AadGroupIds aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee",
+      "commandLineArgs": "--subscriptionid 594bef42-33f3-42df-b056-e7399d6ae7a0 --regionname westus2 --mainidentifierprefix toatest --AadGroupIds 3f16f955-665a-423c-b00b-492ba77d48b8 --HelmBinaryPath %USERPROFILE%\\AppData\\Local\\Microsoft\\WinGet\\Packages\\Helm.Helm_Microsoft.Winget.Source_8wekyb3d8bbwe\\windows-amd64\\helm.exe --EnableIngress true --LetsEncryptEmail sc-vny406119@microsoft.com --DebugLogging true --KubectlBinaryPath %USERPROFILE%\\AppData\\Local\\Microsoft\\WinGet\\Packages\\Kubernetes.kubectl_Microsoft.Winget.Source_8wekyb3d8bbwe\\kubectl.exe",
```
Comment (MattMcL4475) [inline src/deploy-tes-on-azure/Properties/launchSettings.json:5]: this all needs to be removed

File: src/deploy-tes-on-azure/Deployer.cs [hunk 5]
```diff
@@ -76,13 +69,18 @@ public class Deployer(Configuration configuration)
...
-            .WaitAndRetryAsync(3, retryAttempt => System.TimeSpan.FromSeconds(1));
+            .WaitAndRetryAsync(3, retryAttempt => TimeSpan.FromSeconds(1));
 
-        private static readonly System.TimeSpan longRetryWaitTime = System.TimeSpan.FromSeconds(15);
+        private static readonly TimeSpan longRetryWaitTime = TimeSpan.FromSeconds(15);
+
+        /// <summary>
+        /// Grants full access to manage all resources, but does not allow you to assign roles in Azure RBAC, manage assignments in Azure Blueprints, or share image galleries.
+        /// </summary>
+        private static readonly ResourceIdentifier All_Role_Contributor = AuthorizationRoleDefinitionResource.CreateResourceIdentifier(string.Empty, new("b24988ac-6180-42a0-ab88-20f7382dd24c"));
```
Comment (MattMcL4475) [inline src/deploy-tes-on-azure/Deployer.cs:83]: Please include link to where this magic GUID came from

File: src/deploy-tes-on-azure/Deployer.cs [hunk 6]
```diff
@@ -111,22 +109,56 @@ public class Deployer(Configuration configuration)
...
+            ArgumentNullException.ThrowIfNull(GetAsync);
+
+            var result = await GetAsync(cancellationToken);
+            OnAcquisition?.Invoke(result);
+            return result;
+        }
+
+        private Azure.Storage.StorageSharedKeyCredential GetStorageSharedKeyCredential(StorageAccountData storageAccount)
+        {
+            return storageKeys.GetOrAdd(storageAccount.Id, id =>
```
Comment (MattMcL4475) [inline src/deploy-tes-on-azure/Deployer.cs:145]: In the future it is likely that storage keys will need to be disabled by default for storage accounts, is there another way to do this?

File: src/TesApi.Web/Storage/StorageAccessProvider.cs [hunk 7]
```diff
@@ -95,7 +95,15 @@ public async Task<string> DownloadBlobAsync(Uri blobAbsoluteUrl, CancellationTok
     /// <inheritdoc/>
     public async Task<IList<Uri>> GetBlobUrlsAsync(Uri blobVirtualDirectory, CancellationToken cancellationToken)
     {
-        return (await AzureProxy.ListBlobsAsync(blobVirtualDirectory, cancellationToken)).Select(b => b.Uri).ToList();
+        Azure.Storage.Blobs.BlobUriBuilder blobBuilder = new(blobVirtualDirectory) { Sas = null };
+        return (await AzureProxy.ListBlobsAsync(blobVirtualDirectory, cancellationToken)).Select(GetBlobUri).ToList();
+
+        Uri GetBlobUri(Azure.Storage.Blobs.Models.BlobItem blob)
+        {
+            // This implementation assumes there are no parallel invocations in the same parent method call.
```
Comment (MattMcL4475) [inline src/TesApi.Web/Storage/StorageAccessProvider.cs:103]: any concerns about thread safety or this resulting in random bugs?

File: src/TesApi.Web/Management/ArmBatchQuotaProvider.cs [hunk 8]
```diff
@@ -79,33 +78,33 @@ public async Task<BatchVmCoreQuota> GetVmCoreQuotaAsync(bool lowPriority, Cancel
     /// </summary>
     /// <returns></returns>
     public virtual async Task<AzureBatchAccountQuotas> GetBatchAccountQuotasAsync(CancellationToken cancellationToken)
-        => await appCache.GetOrCreateAsync(clientsFactory.BatchAccountInformation.ToString(), _1 => GetBatchAccountQuotasImplAsync(cancellationToken)); // TODO: Consider expiring the quota daily, because quota can be changed.
+        => await appCache.GetOrCreateAsync(clientsFactory.BatchAccountInformation.ToString(), entry =>
+        {
+            entry.AbsoluteExpirationRelativeToNow = TimeSpan.FromDays(1);
```
Comment (MattMcL4475) [inline src/TesApi.Web/Management/ArmBatchQuotaProvider.cs:83]: what does this have to do with "Move away from ADAL"?

File: src/TesApi.Web/BatchScheduler.cs [hunk 9]
```diff
@@ -1118,47 +1135,52 @@ private static BatchModels.BatchPoolIdentity GetBatchPoolIdentity(string[] ident
...
+                Identity = poolIdentity,
+                VmSize = vmSize,
+                ScaleSettings = new() { AutoScale = new(BatchPool.AutoPoolFormula(preemptable, 1)) { EvaluationInterval = BatchPool.AutoScaleEvaluationInterval } },
+                DeploymentConfiguration = new() { VmConfiguration = vmConfig },
                 //ApplicationPackages = ,
                 StartTask = await GetStartTaskAsync(name, vmConfig, vmFamily, cancellationToken),
                 TargetNodeCommunicationMode = BatchModels.NodeCommunicationMode.Simplified,
             };
 
+            poolSpecification.Metadata.Add(new(string.Empty, name));
```
Comment (MattMcL4475) [inline src/TesApi.Web/BatchScheduler.cs:1176]: how is this related to "move away from ADAL"?

File: src/TesApi.Web/BatchScheduler.cs [hunk 10]
```diff
@@ -401,7 +409,7 @@ public async Task UploadTaskRunnerIfNeeded(CancellationToken cancellationToken)
         {
             var blobUri = await storageAccessProvider.GetInternalTesBlobUrlAsync(NodeTaskRunnerFilename, cancellationToken);
             var blobProperties = await azureProxy.GetBlobPropertiesAsync(blobUri, cancellationToken);
-            if (!runnerMD5.Equals(blobProperties?.ContentMD5, StringComparison.OrdinalIgnoreCase))
+            if (!(await File.ReadAllTextAsync(Path.Combine(AppContext.BaseDirectory, $"scripts/{NodeTaskRunnerMD5HashFilename}"), cancellationToken)).Trim().Equals(Convert.ToBase64String(blobProperties?.ContentHash ?? []), StringComparison.OrdinalIgnoreCase))
```
Comment (MattMcL4475) [inline src/TesApi.Web/BatchScheduler.cs:412]: how is this realted to move away from ADAL?

Comment (BMurri) [reply src/TesApi.Web/BatchScheduler.cs:412]: Moving away from ADAL includes moving away from the libraries that depend (directly or indirectly) on ADAL. This is the same functionality in the newer Azure Blob Storage library that works with MSAL (the old one requires ADAL).

File: src/TesApi.Web/BatchScheduler.cs (hunk 9 above)
Comment (BMurri) [reply src/TesApi.Web/BatchScheduler.cs:1176]: The new MSAL-related library does not allow setting the name/id of the pool in the pool data object, so I use a metadata entry with an empty key to pass the name to the code that will submit the creation to the system (whether MSI or Terra) without having to refactor everything in between to add another parameter that wasn't needed before.

File: src/TesApi.Web/Management/ArmBatchQuotaProvider.cs (hunk 8 above)
Comment (BMurri) [reply src/TesApi.Web/Management/ArmBatchQuotaProvider.cs:83]: You got me here. This was a drive-by to get rid of a simple TODO.

File: src/TesApi.Web/Storage/StorageAccessProvider.cs (hunk 7 above)
Comment (BMurri) [reply src/TesApi.Web/Storage/StorageAccessProvider.cs:103]: Not with the current implementation. If this is refactored in the future, there could be a concern. That's why I added this comment.

File: src/deploy-tes-on-azure/Deployer.cs (hunk 5 above)
Comment (BMurri) [reply src/deploy-tes-on-azure/Deployer.cs:83]: Placing link here so that I can add it to the code: https://learn.microsoft.com/en-us/azure/role-based-access-control/built-in-roles#general

Comment (BMurri) [reply src/deploy-tes-on-azure/Deployer.cs:83]: The ADAL-connected libraries provided a property that contained this guid with a nice name. The MSAL-connected libraries do not provide such a convenience.

File: src/deploy-tes-on-azure/Deployer.cs (hunk 6 above)
Comment (BMurri) [reply src/deploy-tes-on-azure/Deployer.cs:145]: Yes, there is, if the calling user has the correct roles in place (being the owner does not provide the required privileges). The calling user in this case is not the managed identity but the deploying agent.

That will need to be a separate PR, probably from an issue we should write that should also encompass CoA as well.

Comment (BMurri) [reply src/deploy-tes-on-azure/Deployer.cs:145]: We might be able to add the role as we go, but I worry about robustness issues cropping up

File: src/deploy-tes-on-azure/Properties/launchSettings.json (hunk 4 above)
Comment (BMurri) [reply src/deploy-tes-on-azure/Properties/launchSettings.json:5]: This file was added to some commit by mistake. I will revert it.

File: src/deploy-tes-on-azure/Utility.cs (hunk 2 above)
Comment (BMurri) [reply src/deploy-tes-on-azure/Utility.cs:156]: I just copied the code from the ADAL-connected library we are retiring. The new MSAL-connected library uses a completely different style of generated names and I didn't feel this is the PR to introduce this kind of behavior change. I didn't question the code since it's functional, but you have a good point. I'll clean this up.

File: src/TesApi.Web/Storage/StorageAccessProvider.cs (hunk 7 above)
Comment (BMurri) [reply src/TesApi.Web/Storage/StorageAccessProvider.cs:103]: I'll update the comment to describe the concern

**Comment by MattMcL4475** [comment]
@BMurri please just update the PR title, description, issue to include the totality of the functional changes this makes (if any).  thanks!

**Review by MattMcL4475 — APPROVED** [review]
@BMurri please just update the PR title, description, issue to include the totality of the functional changes this makes (if any).  thanks!
